corpora/en/wordnet/doc/man/wn.1
'\" t
.\" $Id$
.tr ~
.TH WN 1WN "Dec 2006" "WordNet 3.0" "WordNet\(tm User Commands"
.SH NAME
wn \- command line interface to WordNet lexical database  
.SH SYNOPSIS
\fBwn\fP [ \fIsearchstr\fP ] [ \fB\-h\fP] [ \fB\-g\fP ] [ \fB\-a\fP ] [ \fB\-l\fP ] [ \fB\-o\fP ] [ \fB\-s\fP ] [ \fB\-n\fI#\fR ] [ \fIsearch_option\fP... ]
.SH DESCRIPTION
\fBwn(\|)\fP provides a command line interface to the WordNet
database, allowing synsets and relations to be displayed as formatted
text.  For each word, different searches are provided, based on
syntactic category and pointer types.  Although only base forms of
words are usually stored in WordNet, users may search for inflected
forms.  A morphological process is applied to the search string to
generate a form that is present in WordNet.

The command line interface is often useful when writing scripts to
extract information from the WordNet database.  Post-processing of the
output with various scripting tools can reformat the results as
desired.
.SH OPTIONS
.TP 15
.B \-h
Print help text before search results.
.TP 15
.B \-g
Display textual glosses associated with synsets.
.TP 15
.B \-a
Display lexicographer file information.
.TP 15
.B \-o
Display synset offset of each synset.
.TP 15
.B \-s
Display each word's sense numbers in synsets.
.TP 15
.B \-l
Display the WordNet copyright notice, version number, and license.
.TP 15
.B \-n\fI#\fP
Perform search on sense number \fI#\fP only.
.TP 15
\fB-over\fP
Display overview of all senses of \fIsearchstr\fP in all syntactic
categories. 
.SS Search Options
Note that the last letter of \fIsearch_option\fP generally denotes the
part of speech that the search applies to: \fBn\fP for nouns, \fBv\fP
for verbs, \fBa\fP for adjectives, and \fBr\fP for adverbs.  Multiple
searches may be done for \fIsearchstr\fP with a single command by
specifying all the appropriate search options.

.TP 15
\fB\-syns\fP(\fIn\fP | \fIv\fP | \fIa\fP | \fIr\fP) 
Display synonyms
and immediate hypernyms of synsets containing \fIsearchstr\fP.
Synsets are ordered by estimated frequency of use.  For adjectives, if
\fIsearchstr\fP is in a head synset, the cluster's satellite synsets
are displayed in place of hypernyms.  If \fIsearchstr\fP is in a
satellite synset, its head synset is also displayed.
.TP 15
\fB\-simsv\fP
Display verb synonyms and
immediate hypernyms of synsets containing \fIsearchstr\fP.  Synsets
are grouped by similarity of meaning.
.TP 15
\fB\-ants\fP(\fIn\fP | \fIv\fP | \fIa\fP | \fIr\fP)
Display synsets containing antonyms of \fIsearchstr\fP.
For adjectives, if \fIsearchstr\fP is
in a head synset, \fIsearchstr\fP has a direct antonym.
The head synset for the direct antonym is displayed along
with the direct antonym's satellite synsets.  If \fIsearchstr\fP is in a
satellite synset, \fIsearchstr\fP has an indirect antonym via the
head synset, which is displayed.
.TP 15
\fB\-faml\fP(\fIn\fP | \fIv\fP | \fIa\fP | \fIr\fP)
Display familiarity and polysemy information for \fIsearchstr\fP.
.TP 15
\fB\-hype\fP(\fIn\fP | \fIv\fP)
Recursively display hypernym (superordinate) tree for \fIsearchstr\fP
(\fIsearchstr\fP \fIIS A KIND OF _____\fP relation).
.TP 15
\fB\-hypo\fP(\fIn\fP | \fIv\fP)
Display immediate hyponyms (subordinates) for \fIsearchstr\fP
(\fI_____ IS A KIND OF\fP \fIsearchstr\fP relation).
.TP 15
\fB\-tree\fP(\fIn\fP | \fIv\fP)
Display hyponym (subordinate) tree for \fIsearchstr\fP.  This is
a recursive search that finds the hyponyms of each hyponym.
.TP 15
\fB\-coor\fP(\fIn\fP | \fIv\fP)
Display the coordinates (sisters) of \fIsearchstr\fP.  This
search prints the immediate hypernym for each synset that contains
\fIsearchstr\fP and the hypernym's immediate hyponyms.
.TP 15
\fB\-deri\fP(\fIn\fP | \fIv\fP)
Display derivational morphology links between noun and verb forms.
.TP 15
\fB\-domn\fP(\fIn\fP | \fIv\fP | \fIa\fP | \fIr\fP)
Display domain that \fIsearchstr\fP has been classified in.
.TP 15
\fB\-domt\fP(\fIn\fP | \fIv\fP | \fIa\fP | \fIr\fP)
Display all terms classified as members of the \fIsearchstr\fP's domain.
.TP 15
.B \-subsn
Display substance meronyms of \fIsearchstr\fP
(\fIHAS SUBSTANCE\fP relation).
.TP 15
.B \-partn
Display part meronyms of \fIsearchstr\fP
(\fIHAS PART\fP relation).
.TP 15
.B \-membn
Display member meronyms of \fIsearchstr\fP
(\fIHAS MEMBER\fP relation).
.TP 15
.B \-meron
Display all meronyms of \fIsearchstr\fP
(\fIHAS PART, HAS MEMBER, HAS SUBSTANCE\fP relations).
.TP 15
.B \-hmern
Display meronyms for \fIsearchstr\fP tree.  This is a recursive search
that prints all the meronyms of \fIsearchstr\fP and all of
its hypernyms.
.TP 15
.B \-sprtn
Display \fIpart of\fP holonyms of \fIsearchstr\fP
(\fIPART OF\fP relation).
.TP 15
.B \-smemn
Display \fImember of\fP holonyms of \fIsearchstr\fP
(\fIMEMBER OF\fP relation).
.TP 15
.B \-ssubn
Display \fIsubstance of\fP holonyms of \fIsearchstr\fP
(\fISUBSTANCE OF\fP relation).
.TP 15
.B \-holon
Display all holonyms of \fIsearchstr\fP
(\fIPART OF, MEMBER OF, SUBSTANCE OF\fP relations).
.TP 15
.B \-hholn
Display holonyms for \fIsearchstr\fP tree.  This is a recursive search
that prints all the holonyms of \fIsearchstr\fP and all of each
holonym's holonyms.
.TP 15
.B \-entav
Display entailment relations of \fIsearchstr\fP.
.TP 15
.B \-framv
Display applicable verb sentence frames for \fIsearchstr\fP. 
.TP 15
.B \-causv
Display \fIcause to\fP relations of \fIsearchstr\fP.
.TP 15
\fB \-pert\fP(\fIa\fP | \fIr\fP)
Display pertainyms of \fIsearchstr\fP.
.TP 15
\fB \-attr\fP(\fIn\fP | \fIa\fP)
Display adjective values for noun attribute, or noun attributes of
adjective values.
.TP 15
\fB\-grep\fP(\fIn\fP | \fIv\fP | \fIa\fP | \fIr\fP)
List compound words containing \fIsearchstr\fP as a substring.
.SH SEARCH RESULTS
The results of a search are written to the standard output.  For each
search, the output consists a one line description of the search,
followed by the search results.

All searches other than \fB\-over\fP list all senses matching the
search results in the following general format.  Items enclosed in
italicized square brackets (\fI[~...~]\fP) may not be present.

.RS
One line listing the number of senses matching the search request.

Each sense matching the search requested displayed as follows:

.nf
	\fBSense \fIn\fR
	\fI[\fB{\fIsynset_offset\fB}\fI] [\fB<\fIlex_filename\fB>\fI]~~word1[\fB#\fIsense_number][,~~word2...]\fR
.fi

Where \fIn\fP is the sense number of the search word,
\fIsynset_offset\fP is the byte offset of the synset in the
\fBdata.\fIpos\fR file corresponding to the syntactic category,
\fIlex_filename\fP is the name of the lexicographer file that the
synset comes from, \fIword1\fP is the first word in the synset (note
that this is not necessarily the search word) and \fIsense_number\fP
is the WordNet sense number assigned to the preceding word.
\fIsynset_offset, lex_filename\fP, and \fIsense_number\fP are
generated when the \fB\-o, \-a,\fP and \fB\-s\fP options,
respectively, are specified.

The synsets matching the search requested are printed below each sense's
synset output described above.  Each line of output is preceded by a
marker (usually \fB=>\fP), then a synset, formatted as described
above.  If a search traverses more one level of the tree, then
successive lines are indented by spaces corresponding to its level in
the hierarchy.  When the \fB\-g\fP option is specified, synset glosses
are displayed in parentheses at the end of each synset.  Each synset
is printed on one line.

Senses are generally ordered from most to least frequently used, with
the most common sense numbered \fB1\fP.  Frequency of use is
determined by the number of times a sense is tagged in the various
semantic concordance texts.  Senses that are not semantically tagged
follow the ordered senses.  Note that this ordering is only an
estimate based on usage in a small corpus.

Verb senses can be grouped by similarity of meaning, rather
than ordered by frequency of use.  The \fB\-simsv\fP search prints all
senses that are close in meaning together, with a line of dashes
indicating the end of a group.  See
.BR wngroups (7WN)
for a discussion of how senses are grouped.

The \fB\-over\fP search displays an overview of all the senses of the
search word in all syntactic categories.  The results of this search
are similar to the \fB\-syns\fP search, however no additional
(ex. hypernym) synsets are displayed, and synset glosses are always
printed.  The senses are grouped by syntactic category, and each
synset is annotated as described above with \fIsynset_offset\fP,
\fIlex_filename\fP, and \fIsense_number\fP as dictated by the
\fB\-o, \-a,\fP and \fB\-s\fP options.  The overview search also
indicates how many of the senses in each syntactic category are
represented in the tagged texts.  This is a way for the user to
determine whether a sense's sense number is based on semantic tagging
data, or was arbitrarily assigned.   For each sense that has
appeared in such texts, the number of semantic tags to that sense are
indicated in parentheses after the sense number.

If a search cannot be performed on some senses of \fIsearchstr\fP, the
search results are headed by a string of the form:
.nf
	X of Y senses of \fIsearchstr\fP
.fi

The output of the \fB\-deri\fP search shows word forms that are
morphologically related to \fBsearchstr\fP. Each word form pointed to
from \fIsearchstr\fP is displayed, preceded by \fBRELATED TO->\fP and
the syntactic category of the link, followed, on the next line, by its
synset.  Printed after the word form is \fB#\fP\fIn\fP where \fIn\fP
indicates the WordNet sense number of the term pointed to.

The \fB\-domn\fP and \fB\-domt\fP searches show the domain that a
synset has been classified in and, conversely, all of the terms that
have been assigned to a specific domain.  A domain is
either a \fBTOPIC,\fP \fBREGION\fP or \fBUSAGE,\fP as reflected in
the specific pointer character stored in the database, and displayed
in the output.  A \fB\-domn\fP search on a term shows the domain, if
any, that each synset containing \fIsearchstr\fP has been classified
in.  The output display shows the domain type (\fBTOPIC,\fP
\fBREGION\fP or \fBUSAGE\fP), followed by the syntactic category of
the domain synset and the terms in the synset.  Each term is followed
by \fB#\fP\fIn\fP where \fIn\fP indicates the WordNet sense number of
the term.  The converse search, \fB\-domt\fP, shows all of the synsets
that have been placed into the domain \fIsearchstr\fP, with analogous
markers. 

When \fB\-framv\fP is specified, sample illustrative sentences and
generic sentence frames are displayed.  If a sample sentence is found,
the base form of \fIsearch\fP is substituted into the sentence, and it
is printed below the synset, preceded with the \fBEX:\fP marker.  When
no sample sentences are found, the generic sentence frames are
displayed.  Sentence frames that are acceptable for all words in a
synset are preceded by the marker \fB*>\fP.  If a frame is acceptable
for the search word only, it is preceded by the marker \fB=>\fP.

Search results for adjectives are slightly different from those for
other parts of speech.  When an adjective is printed, its direct
antonym, if it has one, is also printed in parentheses.  When
\fIsearchstr\fP is in a head synset, all of the head synset's
satellites are also displayed.  The position of an adjective in
relation to the noun may be restricted to the \fIprenominal\fP,
\fIpostnominal\fP or \fIpredicative\fP position.  Where present, these
restrictions are noted in parentheses.

When an adjective is a participle of a verb, the output indicates the
verb and displays its synset.

When an adverb is derived from an adjective, the specific adjectival
sense on which it is based is indicated.

The morphological transformations performed by the search code may
result in more than one word to search for.  WordNet automatically
performs the requested search on all of the strings and returns the
results grouped by word.  For example, the verb \fBsaw\fP is both the
present tense of \fBsaw\fP and the past tense of \fBsee\fP.  When
passed \fIsearchstr\fP \fBsaw\fP, WordNet performs the desired search
first on \fBsaw\fP and next on \fBsee\fP, returning the list of
\fBsaw\fP senses and search results, followed by those for \fBsee\fP.
.SH EXIT STATUS
\fBwn(\|)\fP normally exits with the number of senses displayed.  If
\fIsearchword\fP is not found in WordNet, it exits with \fB0\fP.

If the WordNet database cannot be opened, an error messages is
displayed and \fBwn(\|)\fP exits with \fB-1\fP.
.SH ENVIRONMENT VARIABLES (UNIX)
.TP 20
.B WNHOME
Base directory for WordNet.  Default is
\fB/usr/local/WordNet-3.0\fP.
.TP 20
.B WNSEARCHDIR
Directory in which the WordNet database has been installed.  
Default is \fBWNHOME/dict\fP.
.SH REGISTRY (WINDOWS)
.TP 20
.B HKEY_LOCAL_MACHINE\eSOFTWARE\eWordNet\e3.0\eWNHome
Base directory for WordNet.  Default is
\fBC:\eProgram~Files\eWordNet\e3.0\fP.
.SH FILES
.TP 20
.B index.\fIpos\fP
database index files
.TP 20
.B data.\fIpos\fP
database data files
.TP 20
.B *.vrb
files of sentences illustrating the use of verbs
.TP 20
.B \fIpos\fP.exc
morphology exception lists
.SH SEE ALSO
.BR wnintro (1WN),
.BR wnb (1WN),
.BR wnintro (3WN),
.BR lexnames (5WN),
.BR senseidx (5WN)
.BR wndb (5WN), 
.BR wninput (5WN),
.BR morphy (7WN),
.BR wngloss (7WN),
.BR wngroups (7WN).
.SH BUGS
Please report bugs to wordnet@princeton.edu.


corpora/en/wordnet/src/wn.c
/* 

  wn.c - Command line interface to WordNet

*/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "wn.h"

static char *Id = "$Id: wn.c,v 1.13 2005/01/31 19:19:09 wn Rel $";

static struct {
    char *option;		/* user's search request */
    int search;			/* search to pass findtheinfo() */
    int pos;			/* part-of-speech to pass findtheinfo() */
    int helpmsgidx;		/* index into help message table */
    char *label;		/* text for search header message */
} *optptr, optlist[] = {
    { "-synsa", SIMPTR,	ADJ, 0, "Similarity" },
    { "-antsa", ANTPTR,	ADJ, 1, "Antonyms" },
    { "-perta", PERTPTR, ADJ, 0, "Pertainyms" },
    { "-attra", ATTRIBUTE, ADJ, 2, "Attributes" },
    { "-domna", CLASSIFICATION, ADJ, 3, "Domain" },
    { "-domta", CLASS, ADJ, 4, "Domain Terms" },
    { "-famla", FREQ, ADJ, 5, "Familiarity" },
    { "-grepa", WNGREP, ADJ, 6, "Grep" },

    { "-synsn", HYPERPTR, NOUN, 0, "Synonyms/Hypernyms (Ordered by Estimated Frequency)" },
    { "-antsn", ANTPTR,	NOUN, 2, "Antonyms" },
    { "-coorn", COORDS, NOUN, 3, "Coordinate Terms (sisters)" },
    { "-hypen", -HYPERPTR, NOUN, 4, "Synonyms/Hypernyms (Ordered by Estimated Frequency)" },
    { "-hypon", HYPOPTR, NOUN, 5, "Hyponyms" },
    { "-treen", -HYPOPTR, NOUN, 6, "Hyponyms" },
    { "-holon", HOLONYM, NOUN, 7, "Holonyms" },
    { "-sprtn", ISPARTPTR, NOUN, 7, "Part Holonyms" },
    { "-smemn", ISMEMBERPTR, NOUN, 7, "Member Holonyms" },
    { "-ssubn", ISSTUFFPTR, NOUN, 7, "Substance Holonyms" },
    { "-hholn",	-HHOLONYM, NOUN, 8, "Holonyms" },
    { "-meron", MERONYM, NOUN, 9, "Meronyms" },
    { "-subsn", HASSTUFFPTR, NOUN, 9, "Substance Meronyms" },
    { "-partn", HASPARTPTR, NOUN, 9, "Part Meronyms" },
    { "-membn", HASMEMBERPTR, NOUN, 9, "Member Meronyms" },
    { "-hmern", -HMERONYM, NOUN, 10, "Meronyms" },
    { "-nomnn", DERIVATION, NOUN, 11, "Derived Forms" },
    { "-derin", DERIVATION, NOUN, 11, "Derived Forms" },
    { "-domnn", CLASSIFICATION, NOUN, 13, "Domain" },
    { "-domtn", CLASS, NOUN, 14, "Domain Terms" },
    { "-attrn", ATTRIBUTE, NOUN, 12, "Attributes" },
    { "-famln", FREQ, NOUN, 15, "Familiarity" },
    { "-grepn", WNGREP, NOUN, 16, "Grep" },

    { "-synsv", HYPERPTR, VERB, 0, "Synonyms/Hypernyms (Ordered by Estimated Frequency)" },
    { "-simsv", RELATIVES, VERB, 1, "Synonyms (Grouped by Similarity of Meaning)" },
    { "-antsv", ANTPTR, VERB, 2, "Antonyms" },
    { "-coorv", COORDS, VERB, 3, "Coordinate Terms (sisters)" },
    { "-hypev", -HYPERPTR, VERB, 4, "Synonyms/Hypernyms (Ordered by Estimated Frequency)" },
    { "-hypov", HYPOPTR, VERB, 5, "Troponyms (hyponyms)" },
    { "-treev", -HYPOPTR, VERB, 5, "Troponyms (hyponyms)" },
    { "-tropv", -HYPOPTR, VERB, 5, "Troponyms (hyponyms)" },
    { "-entav", ENTAILPTR, VERB, 6, "Entailment" },
    { "-causv", CAUSETO, VERB, 7, "\'Cause To\'" },
    { "-nomnv", DERIVATION, VERB, 8, "Derived Forms" },
    { "-deriv", DERIVATION, VERB, 8, "Derived Forms" },
    { "-domnv", CLASSIFICATION, VERB, 10, "Domain" },
    { "-domtv", CLASS, VERB, 11, "Domain Terms" },
    { "-framv", FRAMES, VERB, 9, "Sample Sentences" },
    { "-famlv", FREQ, VERB, 12, "Familiarity" },
    { "-grepv", WNGREP, VERB, 13, "Grep" },

    { "-synsr", SYNS, ADV, 0, "Synonyms" },
    { "-antsr", ANTPTR, ADV, 1, "Antonyms" },
    { "-pertr", PERTPTR, ADV, 0, "Pertainyms" },
    { "-domnr", CLASSIFICATION, ADV, 2, "Domain" },
    { "-domtr", CLASS, ADV, 3, "Domain Terms" },
    { "-famlr", FREQ, ADV, 4, "Familiarity" },
    { "-grepr", WNGREP, ADV, 5, "Grep" },

    { "-over", OVERVIEW, ALL_POS, -1, "Overview" },
    { NULL, 0, 0, 0, NULL }
};

struct {
    char *template;		/* template for generic search message */
    char *option;		/* text for help message */
    char *helpstr;	
} searchstr[] = {		/* index by search type type */
    { NULL, NULL, NULL },
    { "-ants%c",	"-ants{n|v|a|r}",	"\t\tAntonyms", },
    { "-hype%c",	"-hype{n|v}",	"\t\tHypernyms", },
    { "-hypo%c, -tree%c",	"-hypo{n|v}, -tree{n|v}",
	  "\tHyponyms & Hyponym Tree", },
    { "-enta%c",	"-entav\t",	"\t\tVerb Entailment", },
    { "-syns%c",	"-syns{n|v|a|r}",	"\t\tSynonyms (ordered by estimated frequency)", },
    { "-smem%c",	"-smemn\t",	"\t\tMember of Holonyms", },
    { "-ssub%c",	"-ssubn\t",	"\t\tSubstance of Holonyms", },
    { "-sprt%c",	"-sprtn\t",	"\t\tPart of Holonyms", },
    { "-memb%c",	"-membn\t",	"\t\tHas Member Meronyms", },
    { "-subs%c",	"-subsn\t",	"\t\tHas Substance Meronyms", },
    { "-part%c",	"-partn\t",	"\t\tHas Part Meronyms", },
    { "-mero%c",	"-meron\t",	"\t\tAll Meronyms", },
    { "-holo%c",	"-holon\t",	"\t\tAll Holonyms", },
    { "-caus%c",	"-causv\t",	"\t\tCause to", },
    { NULL, NULL, NULL },	/* PPLPTR - no specific search */
    { NULL, NULL, NULL },	/* SEEALSOPTR - no specific search */
    { "-pert%c",	"-pert{a|r}",	"\t\tPertainyms", },
    { "-attr%c",	"-attr{n|a}",	"\t\tAttributes", },
    { NULL, NULL, NULL },	/* verb groups - no specific pointer */
    { "-deri%c",        "-deri{n|v}",   "\t\tDerived Forms",},
    { "-domn%c", "-domn{n|v|a|r}", "\t\tDomain" },
    { "-domt%c", "-domt{n|v|a|r}", "\t\tDomain Terms" }, 
    { NULL, NULL, NULL },	/* SYNS - taken care of with SIMPTR */
    { "-faml%c",	"-faml{n|v|a|r}",	"\t\tFamiliarity & Polysemy Count", },
    { "-fram%c",	"-framv\t",	"\t\tVerb Frames", },
    { "-coor%c",	"-coor{n|v}",	"\t\tCoordinate Terms (sisters)", },
    { "-sims%c",	"-simsv\t",	"\t\tSynonyms (grouped by similarity of meaning)", },
    { "-hmer%c",	"-hmern\t",	"\t\tHierarchical Meronyms", },
    { "-hhol%c", 	"-hholn\t",	"\t\tHierarchical Holonyms" },
    { "-grep%c",	"-grep{n|v|a|r}", "\t\tList of Compound Words" },
    { "-over", "-over\t", "\t\tOverview of Senses" },
};

static int getoptidx(char *), cmdopt(char *);
static int searchwn(int, char *[]);
static int do_search(char *, int, int, int, char *);
static int do_is_defined(char *);
static void printusage(), printlicense(),
       printsearches(char *, int, unsigned long);
static int error_message(char *);

main(int argc,char *argv[])
{
    display_message = error_message;
    
    if (argc < 2) {
	printusage();
	exit(-1);
    } else if (argc == 2 && !strcmp("-l", argv[1])) {
	printlicense();
	exit(-1);
    }

    if (wninit()) {		/* open database */
	display_message("wn: Fatal error - cannot open WordNet database\n");
	exit (-1);
    }
    
    exit(searchwn(argc, argv));
}

static int searchwn(int ac, char *av[])
{
    int i, j = 1, pos;
    int whichsense = ALLSENSES, help = 0;
    int errcount = 0, outsenses = 0;
    char tmpbuf[256];		/* buffer for constuction error messages */
    
    if (ac == 2)		/* print available searches for word */
	exit(do_is_defined(av[1]));

    /* Parse command line options once and set flags */

    dflag = fileinfoflag = offsetflag = wnsnsflag = 0;

    for(i = 1; i < ac; i++) {
	if(!strcmp("-g",av[i]))
	    dflag++;
	else if (!strcmp("-h",av[i]))
	    help++;
	else if (!strcmp("-l", av[i]))
	    printlicense();
	else if (!strncmp("-n", av[i], 2) && strncmp("-nomn", av[i] ,5))
	    whichsense = atoi(av[i] + 2);
	else if (!strcmp("-a", av[i]))
	    fileinfoflag = 1;
	else if (!strcmp("-o", av[i]))
	    offsetflag = 1;
	else if (!strcmp("-s", av[i]))
	    wnsnsflag = 1;
    }

    /* Replace spaces with underscores before looking in database */

    strtolower(strsubst(av[1], ' ', '_'));

    /* Look at each option in turn.  If it's not a command line option
       (which was processed earlier), perform the search requested. */

    while(av[++j]) {
	if (!cmdopt(av[j])) {	/* not a command line option */
	    if ((i = getoptidx(av[j])) != -1) {
		optptr = &optlist[i];

		/* print help text before search output */
		if (help && optptr->helpmsgidx >= 0)
		    printf("%s\n", helptext[optptr->pos][optptr->helpmsgidx]);

		if (optptr->pos == ALL_POS)
		    for (pos = 1; pos <= NUMPARTS; pos++)
			outsenses += do_search(av[1], pos, optptr->search,
						whichsense, optptr->label);
		else
		    outsenses += do_search(av[1], optptr->pos, optptr->search,
					    whichsense, optptr->label);
	    } else {
		sprintf(tmpbuf, "wn: invalid search option: %s\n", av[j]);
		display_message(tmpbuf);
		errcount++;
	    }
	}
    }
    return(errcount ? -errcount : outsenses);
}

static int do_search(char *searchword, int pos, int search, int whichsense,
		     char *label)
{
    int totsenses = 0;
    char *morphword, *outbuf;

    outbuf = findtheinfo(searchword, pos, search, whichsense);
    totsenses += wnresults.printcnt;
    if (strlen(outbuf) > 0)
	printf("\n%s of %s %s\n%s",
	       label, partnames[pos], searchword, outbuf);

    if (morphword = morphstr(searchword, pos))
	do {
	    outbuf = findtheinfo(morphword, pos, search, whichsense);
	    totsenses += wnresults.printcnt;
	    if (strlen(outbuf) > 0) 
		printf("\n%s of %s %s\n%s",
		       label, partnames[pos], morphword, outbuf);
	} while (morphword = morphstr(NULL, pos));

    return(totsenses);
}

static int do_is_defined(char *searchword)
{
    int i, found = 0;
    unsigned int search;
    char *morphword;

    if (searchword[0] == '-') {
	display_message("wn: invalid search word\n");
	return(-1);
    }
    
    /* Print all valid searches for word in all parts of speech */

    strtolower(strsubst(searchword, ' ', '_'));
	    
    for (i = 1; i <= NUMPARTS; i++) {
	if ((search = is_defined(searchword, i)) != 0) {
	    printsearches(searchword, i, search);
	    found = 1;
	} else
	    printf("\nNo information available for %s %s\n",
		   partnames[i], searchword);

	if ((morphword = morphstr(searchword, i)) != NULL)
	    do {
		if ((search = is_defined(morphword, i)) != 0) {
		    printsearches(morphword, i, search);
		    found = 1;
		} else
		    printf("\nNo information available for %s %s\n",
			   partnames[i], morphword);
	    } while ((morphword = morphstr(NULL, i)) != NULL );
    }
    return(found);
}

static void printsearches(char *word, int dbase, unsigned long search)
{
    int j;

    printf("\nInformation available for %s %s\n", partnames[dbase], word);
    for (j = 1; j <= MAXSEARCH; j++)
	if ((search & bit(j)) && searchstr[j].option) {
	    printf("\t");
	    printf(searchstr[j].template,
		   partchars[dbase], partchars[dbase]);
	    printf(searchstr[j].helpstr);
	    printf("\n");
	}
}

static void printusage()
{
    int i;

    fprintf(stdout,
	    "\nusage: wn word [-hgla] [-n#] -searchtype [-searchtype...]\n");
    fprintf(stdout, "       wn [-l]\n\n");
    fprintf(stdout, "\t-h\t\tDisplay help text before search output\n");
    fprintf(stdout, "\t-g\t\tDisplay gloss\n");
    fprintf(stdout, "\t-l\t\tDisplay license and copyright notice\n");
    fprintf(stdout, "\t-a\t\tDisplay lexicographer file information\n");
    fprintf(stdout, "\t-o\t\tDisplay synset offset\n");
    fprintf(stdout, "\t-s\t\tDisplay sense numbers in synsets\n");
    fprintf(stdout, "\t-n#\t\tSearch only sense number #\n");
    fprintf(stdout,"\nsearchtype is at least one of the following:\n");

    for (i = 1; i <= OVERVIEW; i++)
	if (searchstr[i].option)
	    fprintf(stdout, "\t%s%s\n",
		    searchstr[i].option, searchstr[i].helpstr);
}

static void printlicense()
{
    printf("WordNet Release %s\n\n%s", wnrelease, license);
}

static int cmdopt(char *str)
{

    if (!strcmp("-g", str) ||
	!strcmp("-h", str) ||
	!strcmp("-o", str) ||
	!strcmp("-l", str) ||
	!strcmp("-a", str) ||
	!strcmp("-s", str) ||
	(!strncmp("-n", str, 2) && strncmp("-nomn", str,5)))

	return (1);
    else
	return(0);
}

static int getoptidx(char *searchtype)
{
    int i;

    for (i = 0; optlist[i].option; i++)
	if (!strcmp(optlist[i].option, searchtype))
	    return(i);

    return(-1);
}

static int error_message(char *msg)
{
    fprintf(stderr, msg);
    return(0);
}

/*
  Revision log: (since version 1.5)
  
  $Log: wn.c,v $
  Revision 1.13  2005/01/31 19:19:09  wn
  removed include for license.h

  Revision 1.12  2005/01/27 17:32:37  wn
  removed wnhelp.h

  Revision 1.11  2004/10/25 16:34:43  wn
  removed 1.6 references

  Revision 1.10  2003/07/15 16:50:53  wn
  added domain and domain term searches

  Revision 1.9  2003/07/15 15:53:05  wn
  updated search numbers

  Revision 1.8  2002/03/07 17:52:49  wn
  fixes for 1.7.1

  Revision 1.7  2001/11/06 18:51:59  wn
  added CLASSIFICATION placeholders

  Revision 1.6  2001/10/25 16:56:11  wn
  changed text on synsnym searches to say 'estimated frequency'

  Revision 1.5  2001/07/20 18:15:10  wn
  changed Nominalizations to Derived Forms

  Revision 1.4  2001/06/19 15:06:17  wn
  changed search from long to int

  Revision 1.3  2001/03/30 17:16:05  wn
  cleanups for 1.7

  Revision 1.2  2000/10/30 19:06:49  wn
  added code to handle nominalizations

  Revision 1.1  1998/05/06 18:22:57  wn
  Initial revision

 * Revision 1.58  1997/11/21  19:01:17  wn
 * added simsv search
 *
 * Revision 1.57  1997/09/02  17:10:32  wn
 * changed includes
 *
 * Revision 1.56  1997/08/29  18:43:37  wn
 * rearranged functions
 *
 * Revision 1.55  1997/08/29  16:44:24  wn
 * added code to exit with total senses printed
 *
 * Revision 1.54  1997/08/26  20:29:38  wn
 * added -s option, reorganized code
 *
 * Revision 1.53  1997/08/08  19:19:03  wn
 * major cleanup
 *
 * Revision 1.52  1997/08/05  20:15:31  wn
 * removed WNDEBUG, cleanups
 *
 * Revision 1.51  1995/06/30  19:25:02  wn
 * access first element of OutSenseCount array
 *
  * 
  * Revision 1.1  91/09/17  15:51:09  wn
  * Initial revision
  * 
  */

